README.md
# known-to-hugo
a little tool to help migrate your [Known](https://withknown.com/) website to [Hugo](https://gohugo.io/)

[![Build Status](https://travis-ci.org/nekr0z/known-to-hugo.svg?branch=master)](https://travis-ci.org/nekr0z/known-to-hugo) [![codecov](https://codecov.io/gh/nekr0z/known-to-hugo/branch/master/graph/badge.svg)](https://codecov.io/gh/nekr0z/known-to-hugo) [![Go Report Card](https://goreportcard.com/badge/github.com/nekr0z/known-to-hugo)](https://goreportcard.com/report/github.com/nekr0z/known-to-hugo)

##### Table of Contents
* [Why](#why)
* [How to use](#how)
  * [Options](#command-line-options)
* [Development](#development)
* [Credits](#credits)

#### Help `known-to-hugo` get better!
Join the [development](#development) (or just [buy me a coffee](https://www.buymeacoffee.com/nekr0z), that helps, too).

## Why
[Known](https://withknown.com/) is great and has brought a lot of people to [IndieWeb](https://indieweb.org/). However, its export features are incomplete and have bugs. If you want to start using [Hugo](https://gohugo.io/) instead, you need to get all your content from the Known instance and save it so that Hugo can work with it (a simple MySQL dump wouldn't do). This tool here does exactly that.

## How
Change the theme on your Known website to the builtin "Solo" theme (it may work with other themes, but no one has tested it yet). While your Known site is still up and running, open your command line and run the tool. It will try to save all your posts neatly to Hugo-compatible markdown files. It will also generate a JSON file with webmentions for every page that has them, as well as try to download the images.

**Beware:** the files will be overwritten without asking!

### Command line options
```
-w [website]
```
the domain to look for, as in `evgenykuznetsov.org` or `some-user.withknown.com` or `https://my-awesome-known-site.info`.

```
-ww [path]
```
the part of website to try and download. Default is `/content/posts`, so only posts get downloaded. You can use `-ww ""` to download whatever content is listed on your home page, or `-ww /content/all` to try to get everything there is.

```
-p [directory]
```
the directory to save everything to. Default is `known_website` in your current directory.

```
-d
```
mark all the downloaded pages as drafts for Hugo.

```
-c [number]
```
number of pages to try to process simultaneously. Your server that runs Known might not like `known-to-hugo`'s attempt to download all the posts simultaneously (for example, the [DreamHost](https://www.dreamhost.com/) shared web hosting I use starts serving `503`s instead of pages when I try about 20 processes in parallel), so this option limits the number of pages processed in parallel. Default is `15`.

### Local backups processing
If you happen to have a local backup of your old blog, these are some experimental options for you:
```
-dir [path]
```
tells `known-to-hugo` to work with a local directory rather than a running website. The `-w` and `-ww` options are ignored in this case; site type (below) must be specified:
```
-type gplus
```
the local backup is Google Plus posts directory (as composed with Google Takeout), or
```
-type lj_backup
```
the local backup is a LiveJournal backup made with `ljArchive`, or
```
-type diary.ru
```
the local backup is a whole-site `wget` copy of a diary.ru-hosted blog.

## Development
Pull requests are always welcome!

## Credits
This software includes the following software or parts thereof:
* [goquery](https://github.com/PuerkitoBio/goquery) Copyright © 2012-2016, Martin Angers & Contributors
* [BurntSushi/toml](https://github.com/BurntSushi/toml) Copyright © 2013 TOML authors
* [html-to-markdown](https://github.com/JohannesKaufmann/html-to-markdown) Copyright © 2018 Johannes Kaufmann
* [cascadia](https://github.com/andybalholm/cascadia) Copyright © 2011 Andy Balholm. All rights reserved.
* [The Go Programming Language](https://golang.org) Copyright © 2009 The Go Authors


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.1] - 2019-04-18
### Added
- processing local backups (G+, LJ-backup, diary.ru)

## 0.1.0 - 2020-03-29
*initial release*

[Unreleased]: https://github.com/nekr0z/known-to-hugo/compare/v0.1.1...HEAD
[0.1.1]: https://github.com/nekr0z/known-to-hugo/compare/v0.1.0...v0.1.1


html_test.go
// Copyright (C) 2020 Evgeny Kuznetsov (evgeny@kuznetsov.md)
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

package main

import (
	"path/filepath"
	"testing"

	"golang.org/x/text/encoding/htmlindex"
)

func TestHugo(t *testing.T) {
	tests := map[string]struct {
		file      string
		processor string
		want      string
	}{
		"diary_comments": {"diary_comments.htm", "diary", "diary_comments.md"},
		"diary_pic":      {"diary_pic.htm", "diary", "diary_pic.md"},
		"ljbackup":       {"ljbackup.html", "ljbackup", "ljbackup.md"},
		"ljb_formatting": {"ljb_f.html", "ljbackup", "ljb_f.md"},
		"gp_likes":       {"gp1.html", "gplus", "gp1.md"},
		"gp_comment":     {"gp2.html", "gplus", "gp2.md"},
	}

	for name, tc := range tests {
		t.Run(name, func(t *testing.T) {
			s, err := loadHtmlFile(filepath.Join("testdata", tc.file))
			if err != nil {
				t.Fatal(err)
			}
			var p page
			switch tc.processor {
			case "diary":
				p = diaryPage{s}
			case "ljbackup":
				p = ljbPage{s}
			case "gplus":
				p = gpPage{s}
			default:
				t.Fatal("not implemented")
			}
			g := filepath.Join("testdata", tc.want)
			got := hugo(p, false)
			assertGolden(t, got, g)
		})
	}
}

func TestCanonicalUrl(t *testing.T) {
	tests := map[string]struct {
		file      string
		processor string
		want      string
	}{
		"diary_comments": {"diary_comments.htm", "diary", "p706232_soundcheck.htm"},
		"diary_pic":      {"diary_pic.htm", "diary", "p1225644_nachdenklichkeit.htm"},
		"ljbackup":       {"ljbackup.html", "ljbackup", "170041.html"},
		"ljb_strange":    {"ljb_strange.html", "ljbackup", "69149.html"},
		"gplus":          {"gp1.html", "gplus", "PGs99nrjdi6"},
		"gplus_strange":  {"gp3.html", "gplus", "Cayiw7vFVQK"},
	}

	for name, tc := range tests {
		t.Run(name, func(t *testing.T) {
			s, err := loadHtmlFile(filepath.Join("testdata", tc.file))
			if err != nil {
				t.Fatal(err)
			}
			var p page
			switch tc.processor {
			case "diary":
				p = diaryPage{s}
			case "ljbackup":
				p = ljbPage{s}
			case "gplus":
				p = gpPage{s}
			default:
				t.Fatal("not implemented")
			}
			got := p.canonicalUrl()
			if tc.want != got {
				t.Fatalf("want %s, got %s\n", tc.want, got)
			}
		})
	}
}

func TestProcessImages(t *testing.T) {
	tests := map[string]struct {
		file      string
		processor string
		want      map[string]string
	}{
		"diary_comments": {"diary_comments.htm", "diary", map[string]string{}},
		"diary_pic":      {"diary_pic.htm", "diary", map[string]string{"image0": "https://secure.diary.ru/userdir/4/9/6/1/4961/95265.jpg"}},
	}

	for name, tc := range tests {
		t.Run(name, func(t *testing.T) {
			s, err := loadHtmlFile(filepath.Join("testdata", tc.file))
			if err != nil {
				t.Fatal(err)
			}
			var p page
			switch tc.processor {
			case "diary":
				p = diaryPage{s}
			case "ljbackup":
				p = ljbPage{s}
			default:
				t.Fatal("not implemented")
			}
			cnt := p.content()
			got := cnt.processImages()

			if len(got) != len(tc.want) {
				t.Fatalf("want %s, got %s\n", tc.want, got)
			}

			for k, v := range tc.want {
				if v != got[k] {
					t.Fatalf("want %s, got %s\n", tc.want, got)
				}
			}
		})
	}
}

func TestWebmentions(t *testing.T) {
	tests := map[string]struct {
		file      string
		processor string
		want      string
	}{
		"diary_comments": {"diary_comments.htm", "diary", "diary_comments.json"},
		"diary_pic":      {"diary_pic.htm", "diary", "diary_pic.json"},
		"ljbackup":       {"ljbackup.html", "ljbackup", "ljbackup.json"},
		"gp_likes":       {"gp1.html", "gplus", "gp1.json"},
		"gp_comment":     {"gp2.html", "gplus", "gp2.json"},
	}

	for name, tc := range tests {
		t.Run(name, func(t *testing.T) {
			s, err := loadHtmlFile(filepath.Join("testdata", tc.file))
			if err != nil {
				t.Fatal(err)
			}
			var p page
			switch tc.processor {
			case "diary":
				p = diaryPage{s}
			case "ljbackup":
				p = ljbPage{s}
			case "gplus":
				p = gpPage{s}
			default:
				t.Fatal("not implemented")
			}
			g := filepath.Join("testdata", tc.want)
			got := p.webmentions()
			assertGolden(t, got, g)
		})
	}
}

func TestGetEncoding(t *testing.T) {
	tests := map[string]struct {
		file string
		want string
	}{
		"known": {filepath.Join("testdata", "eter.html"), "utf-8"},
		"diary": {filepath.Join("testdata", "diary_comments.htm"), "windows-1251"},
	}

	for name, tc := range tests {
		t.Run(name, func(t *testing.T) {
			s := loadHtml(t, tc.file)
			e := getEncoding(s)
			got, _ := htmlindex.Name(e)
			if tc.want != got {
				t.Fatalf("want %s, got %s\n", tc.want, got)
			}
		})
	}
}


main_test.go
// Copyright (C) 2020 Evgeny Kuznetsov (evgeny@kuznetsov.md)
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

package main

import (
	"bytes"
	"flag"
	"io/ioutil"
	"os"
	"path/filepath"
	"testing"

	"github.com/PuerkitoBio/goquery"
)

var (
	update = flag.Bool("update", false, "update .golden files")
)

func TestGetTitle(t *testing.T) {
	s := loadHtml(t, filepath.Join("testdata", "tired.html"))
	got := getTitle(s)
	want := "Двигаться дальше…"
	assertString(t, want, got)
}

func TestGetPermalink(t *testing.T) {
	s := loadHtml(t, filepath.Join("testdata", "tired.html"))
	got := getPermalink(s)
	want := "https://evgenykuznetsov.org/2020/двигаться-дальше"
	assertString(t, want, got)
}

func TestGetLikeOf(t *testing.T) {
	s := loadHtml(t, filepath.Join("testdata", "whatever.html"))
	got := getLikeOf(s)
	want := "https://habr.com/ru/post/491672/"
	assertString(t, want, got)
}

func TestGetDtPublished(t *testing.T) {
	s := loadHtml(t, filepath.Join("testdata", "tired.html"))
	got := getDtPublished(s)
	want := "2020-03-17T19:58:16+0000"
	assertString(t, want, got)
}

func TestGetWebmentions(t *testing.T) {
	s := loadHtml(t, filepath.Join("testdata", "eter.html"))
	g := filepath.Join("testdata", "eter.json")
	got, _ := getWebmentions(s)
	assertGolden(t, got, g)
}

func assertGolden(t *testing.T, actual []byte, golden string) {
	t.Helper()

	if *update {
		if _, err := os.Stat(golden); os.IsNotExist(err) {
			if err := ioutil.WriteFile(golden, actual, 0644); err != nil {
				t.Fatal(err)
			}
		} else {
			t.Log("file", golden, "exists, remove it to record new golden result")
		}
	}
	expected, err := ioutil.ReadFile(golden)
	if err != nil {
		t.Error("no file:", golden)
	}

	if !bytes.Equal(actual, expected) {
		t.Fatalf("want:\n%s\ngot:\n%s\n", expected, actual)
	}
}

func assertString(t *testing.T, want, got string) {
	t.Helper()
	if got != want {
		t.Fatalf("want: %v\ngot: %v", want, got)
	}
}

func loadHtml(t *testing.T, path string) *goquery.Selection {
	t.Helper()

	f, err := os.Open(path)
	if err != nil {
		t.Fatal(err)
	}
	defer f.Close()

	doc, err := goquery.NewDocumentFromReader(f)
	if err != nil {
		t.Fatal(err)
	}

	return doc.Find("html")
}


diary.go
// Copyright (C) 2020 Evgeny Kuznetsov (evgeny@kuznetsov.md)
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

package main

import (
	"encoding/json"
	"net/url"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
)

type diaryPage struct {
	*goquery.Selection
}

type diaryComment struct {
	*goquery.Selection
}

func (p diaryPage) canonicalUrl() string {
	u, _ := p.Find(".singlePost").Find(".urlLink").Find("a").Attr("href")
	uri, err := url.Parse(u)
	if err != nil {
		return ""
	}
	return strings.TrimPrefix(uri.Path, "/")
}

func (p diaryPage) title() string {
	return p.Find(".postTitle").Find("h1").Text()
}

func (p diaryPage) date() time.Time {
	s := p.Find(".singlePost")
	date := s.Find(".postDate").Text()
	d := strings.Split(date, ", ")
	if len(d) != 2 {
		return time.Now()
	}
	date = d[1]
	mes := map[string]string{
		"января":   "Jan",
		"февраля":  "Feb",
		"марта":    "Mar",
		"апреля":   "Apr",
		"мая":      "May",
		"июня":     "Jun",
		"июля":     "Jul",
		"августа":  "Aug",
		"сентября": "Sep",
		"октября":  "Oct",
		"ноября":   "Nov",
		"декабря":  "Dec",
	}
	for mr, me := range mes {
		date = strings.ReplaceAll(date, mr, me)
	}

	t := s.Find(".postTitle").Find("span").Text()

	dt, _ := time.ParseInLocation("2 Jan 2006 15:04", date+" "+t, time.Local)
	return dt
}

func (p diaryPage) content() pageContent {
	s := p.Find(".singlePost").Find(".postInner")
	return pageContent{s}
}

func (p diaryPage) tags() []string {
	return nil
}

func (p diaryPage) webmentions() []byte {
	var mentions = struct {
		Type     string    `json:"type"`
		Name     string    `json:"name"`
		Children []mention `json:"children,omitempty"`
	}{Type: "feed", Name: "Webmentions"}

	p.Find(".singleComment").Each(func(i int, s *goquery.Selection) {
		cmt := diaryComment{s}
		m := getWebmention(cmt)
		mentions.Children = append(mentions.Children, m)
	})
	if len(mentions.Children) > 0 {
		b, err := json.MarshalIndent(mentions, "", " ")
		if err != nil {
			panic(err)
		}
		return b
	}
	return nil
}

func (dc diaryComment) author() author {
	n := dc.Find(".authorName").Text()
	p, _ := dc.Find(".commentAuthor").Find("img").Attr("src")
	return author{"card", n, "", p}
}

func (dc diaryComment) content() content {
	c := dc.Find(".postInner")
	t := strings.TrimSpace(c.Text())
	h, _ := c.Html()
	h = strings.TrimSpace(h)
	return content{t, h}
}

func (dc diaryComment) date() string {
	d := dc.Find(".postTitle").Find("span").Text()
	date, _ := time.ParseInLocation("2006-01-02 в 15:04", d, time.Local)
	return date.String()
}

func (dc diaryComment) url() string {
	return ""
}


main.go
// Copyright (C) 2020 Evgeny Kuznetsov (evgeny@kuznetsov.md)
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

//go:generate go run version_generate.go

package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/BurntSushi/toml"
	"github.com/JohannesKaufmann/html-to-markdown"
	"github.com/PuerkitoBio/goquery"
)

var (
	outputDir, website, what, inputDir, siteType string
	concurrency                                  int
	draft                                        bool
)

var version string = "custom"

const frontMatterSeparator = "+++\n"

type content struct {
	Text string `json:"text,omitempty"`
	Html string `json:"html,omitempty"`
}
type author struct {
	Type  string `json:"type,omitempty"`
	Name  string `json:"name,omitempty"`
	Url   string `json:"url,omitempty"`
	Photo string `json:"photo,omitempty"`
}
type mention struct {
	Type     string  `json:"type,omitempty"`
	Property string  `json:"wm-property,omitempty"`
	Author   author  `json:"author"`
	Url      string  `json:"url,omitempty"`
	Date     string  `json:"wm-received"`
	Content  content `json:"content,omitempty"`
}

func main() {
	fmt.Printf("known-to-hugo version %s\n", version)
	flag.BoolVar(&draft, "d", false, "mark each entry as draft")
	flag.IntVar(&concurrency, "c", 15, "number of pages to process simultaneously")
	flag.StringVar(&website, "w", "example.site", "website to scrape")
	flag.StringVar(&outputDir, "p", "./known_website", "directory to save the results to")
	flag.StringVar(&what, "ww", "/content/posts", "section of the site to scrape, use \"\" for default content)")
	flag.StringVar(&inputDir, "dir", "", "input directory")
	flag.StringVar(&siteType, "type", "", "kind of website")
	flag.Parse()
	if !strings.HasPrefix(website, "http://") && !strings.HasPrefix(website, "https://") {
		website = "http://" + website
	}

	if inputDir != "" {
		processDirectory()
	} else {

		pages := getPostLinks(website + what)
		defImg := getDefaultImage(website)
		processPages(pages, defImg)
	}
	fmt.Println("all done!")
}

func processDirectory() {
	switch siteType {
	case "diary.ru":
		blogDir(inputDir, outputDir, "diary")
	case "lj_backup":
		blogDir(inputDir, outputDir, "ljbackup")
	case "gplus":
		blogDir(inputDir, outputDir, "gplus")
	default:
		fmt.Println("not implemented")
	}
}

func processPages(pages []string, defaultImage string) {
	var errors []error
	errC := make(chan error)
	counter := 0
	sem := make(chan struct{}, concurrency)
	for _, page := range pages {
		counter = counter + 1
		go processPage(sem, errC, page, defaultImage)
	}
	for i := 0; i < counter; i++ {
		err := <-errC
		if err != nil {
			errors = append(errors, err)
		}
	}
	if len(errors) > 0 {
		fmt.Printf("the following %v errors occurred:\n", len(errors))
		for _, err := range errors {
			fmt.Println(err)
		}
	}
}

func processPage(sem chan struct{}, errC chan error, url, defaultImage string) {
	sem <- struct{}{}
	defer func() { <-sem }()

	fmt.Printf("processing %s\n", url)
	d, err := getPage(url)
	if err != nil {
		err := fmt.Errorf("could not process %s - %w", url, err)
		errC <- err
		return
	}
	sel := d.Find("html")
	year := getPostYear(sel)
	slug := getPostSlug(url, year)
	dir := filepath.Join(outputDir, year, slug)
	if err := os.MkdirAll(dir, 0755); err != nil {
		panic(err)
	}
	processWebmentions(sel, dir)
	processImages(sel, dir)
	processLinksToFiles(sel, dir)
	processLinksToOwnSite(sel)
	b := parsePage(sel, defaultImage)
	fn := filepath.Join(dir, "index.md")
	if err := ioutil.WriteFile(fn, b, 0644); err != nil {
		panic(err)
	}
	errC <- nil
}

func processWebmentions(sel *goquery.Selection, path string) {
	if b, ok := getWebmentions(sel); ok {
		fn := filepath.Join(path, "webmentions.json")
		if err := ioutil.WriteFile(fn, b, 0644); err != nil {
			panic(err)
		}
	}
}

func getWebmentions(sel *goquery.Selection) ([]byte, bool) {
	var mentions = struct {
		Type     string    `json:"type"`
		Name     string    `json:"name"`
		Children []mention `json:"children,omitempty"`
	}{Type: "feed", Name: "Webmentions"}

	sel.Find(".annotations").Find(".idno-annotation").Each(func(i int, s *goquery.Selection) {
		m := getMention(s)
		mentions.Children = append(mentions.Children, m)
	})
	if len(mentions.Children) > 0 {
		b, err := json.MarshalIndent(mentions, "", " ")
		if err != nil {
			panic(err)
		}
		return b, true
	}
	return nil, false
}

func getMention(sel *goquery.Selection) mention {
	var m = mention{
		Type:   "entry",
		Author: getMentionAuthor(sel),
	}
	if c, ok := getMentionContent(sel); ok {
		m.Content = c
	}
	if t, ok := getMentionType(sel); ok {
		m.Property = t
	}
	m.Url, m.Date = getMentionSource(sel)
	return m
}

func getMentionType(sel *goquery.Selection) (string, bool) {
	s := sel.Find(".idno-annotation-content").Find("p").Eq(0)
	if s.Parent().Is(".e-content") {
		return "comment", false
	}
	a := s.Find("a").Eq(1).Text()
	if strings.HasPrefix(a, "reshared") {
		return "repost-of", true
	}
	a = s.Text()
	if strings.HasSuffix(strings.TrimSpace(a), "liked this post") {
		return "like-of", true
	}
	return a, true
}

func getMentionSource(sel *goquery.Selection) (url, date string) {
	s := sel.Find(".idno-annotation-content").Find("a").Eq(-2)
	d := s.Text()
	dt, err := time.Parse("Jan 02 2006", d)
	if err != nil {
		date = d
	} else {
		date = dt.Format("2006-01-02")
	}
	url, _ = s.Attr("href")
	return
}

func getMentionAuthor(sel *goquery.Selection) author {
	p, _ := sel.Find(".idno-annotation-image").Find("img").Attr("src")
	au := sel.Find(".idno-annotation-content").Find("a").Eq(0)
	n := au.Text()
	u, _ := au.Attr("href")
	s := sel.Find(".h-card")
	if s.Is(".h-card") {
		n = s.Find(".p-name").Text()
		u, _ = s.Find(".p-name").Attr("href")
		p, _ = s.Find(".u-photo").Attr("href")
	}
	return author{"card", n, u, p}
}

func getMentionContent(sel *goquery.Selection) (content, bool) {
	cont := sel.Find(".e-content")
	if cont.Is(".e-content") {
		text := cont.Text()
		html, _ := cont.Html()
		return content{text, html}, true
	}
	return content{}, false
}

func processLinksToOwnSite(sel *goquery.Selection) {
	prefix := strings.TrimSuffix(website, "/")
	sel.Find(".e-content").Find("a").Each(func(i int, s *goquery.Selection) {
		link, _ := s.Attr("href")
		rel := strings.TrimPrefix(link, prefix)
		s.SetAttr("href", rel)
	})
}

func processLinksToFiles(sel *goquery.Selection, dir string) {
	fPrefix := strings.TrimSuffix(website, "/") + "/file/"
	se := sel.Find(".e-content")
	se.Find("a").Each(func(i int, s *goquery.Selection) {
		link, _ := s.Attr("href")
		if strings.HasPrefix(link, fPrefix) {
			pts := strings.Split(link, "/")
			fn := strconv.Itoa(i) + pts[len(pts)-1]
			filename := filepath.Join(dir, fn)
			if err := downloadFile(filename, link); err != nil {
				fmt.Printf("failed to fetch asset: %s - %v", link, err)
				return
			}
			changeHrefs(se, link, fn)
		}
	})
}

func processImages(sel *goquery.Selection, dir string) {
	featured := getFeaturedImage(sel)
	se := sel.Find(".e-content")
	se.Find("img").Each(func(i int, s *goquery.Selection) {
		link, _ := s.Attr("src")
		photoUrl := strings.TrimSuffix(link, "/thumb.jpg")
		fn := "image" + strconv.Itoa(i)
		filename := filepath.Join(dir, fn)
		if err := downloadFile(filename, photoUrl); err != nil {
			fmt.Printf("failed to fetch image: %s - %v", photoUrl, err)
			return
		}

		// fix hrefs
		changeHrefs(se, link, fn)

		// fix featured image, too
		if link == featured {
			sel.Find("meta").Each(func(i int, s *goquery.Selection) {
				v, _ := s.Attr("property")
				if v == "og:image" {
					s.SetAttr("content", fn)
				}
			})
		}
		s.SetAttr("src", fn)
	})
}

func changeHrefs(se *goquery.Selection, link, fn string) {
	se.Find("a").Each(func(i int, s *goquery.Selection) {
		href, _ := s.Attr("href")
		if href == link {
			s.SetAttr("href", fn)
		}
	})
}

func getPostSlug(uri, year string) string {
	u, err := url.Parse(uri)
	if err != nil {
		// Known is buggy as hell
		uf := fixURL(uri)
		u, err = url.Parse(uf)
		if err != nil {
			panic(err)
		}
	}
	slug := strings.TrimPrefix(u.Path, "/"+year+"/")
	return slug
}

func getPostYear(sel *goquery.Selection) string {
	dateString := getDtPublished(sel)
	date, err := time.Parse("2006-01-02T15:04:05-0700", dateString)
	if err != nil {
		panic(err)
	}
	return date.Format("2006")
}

func getPostLinks(url string) []string {
	var links []string
	next := true
	for next {
		fmt.Printf("Processing %s\n", url)
		d, err := getPage(url)
		if err != nil {
			break
		}
		s := d.Find("html")
		var nurl string
		nurl, next = s.Find(".older").Find("a").Attr("href")
		// Known is buggy as hell:
		if nurl == url {
			next = false
		}
		url = nurl
		links = append(links, getPostLinksFromPage(s)...)
	}
	return links
}

func getPostLinksFromPage(sel *goquery.Selection) []string {
	var links []string
	sel.Find(".idno-entry").Each(func(i int, s *goquery.Selection) {
		link := getPermalink(s)
		links = append(links, link)
	})
	return links
}

func parsePage(sel *goquery.Selection, defaultImage string) []byte {
	var b []byte
	b = append(b, getFrontMatter(sel, defaultImage)...)
	b = append(b, []byte(getMd(sel))...)
	return b
}

func getFrontMatter(sel *goquery.Selection, defaultImage string) []byte {
	featured := getFeaturedImage(sel)
	if featured == defaultImage {
		featured = ""
	}

	var frontMatter = map[string]interface{}{
		"title":          getTitle(sel),
		"aliases":        []string{getRelPermalink(sel)},
		"date":           getDtPublished(sel),
		"featured_image": featured,
		"tags":           getTags(sel),
		"reply_to":       getInReply(sel),
		"posse":          getSyndications(sel),
		"like_of":        getLikeOf(sel),
		"draft":          draft,
	}
	buf := new(bytes.Buffer)
	if err := toml.NewEncoder(buf).Encode(frontMatter); err != nil {
		panic(err)
	}
	var b []byte
	b = append(b, []byte(frontMatterSeparator)...)
	b = append(b, buf.Bytes()...)
	b = append(b, []byte(frontMatterSeparator)...)
	return b
}

func getMd(sel *goquery.Selection) string {
	c := sel.Clone()
	converter := md.NewConverter("", true, nil)
	c.Find(".annotations").Remove()
	c.Find(".p-category").Remove()
	got := converter.Convert(c.Find(".e-content"))
	return got
}

func getInReply(sel *goquery.Selection) []string {
	var irt []string
	sel.Find(".u-in-reply-to").Each(func(i int, s *goquery.Selection) {
		rep, _ := s.Attr("href")
		irt = append(irt, rep)
	})
	return irt
}

func getLikeOf(sel *goquery.Selection) string {
	s := sel.Find(".u-like-of")
	// Known is awesome :/
	var like string
	if l, ok := s.Attr("href"); ok {
		if l != "" {
			like = l
		} else {
			like, _ = sel.Find(".unfurl").Attr("data-url")
		}
	}
	return like
}

func getSyndications(sel *goquery.Selection) []string {
	var irt []string
	sel.Find(".u-syndication").Each(func(i int, s *goquery.Selection) {
		rep, _ := s.Attr("href")
		irt = append(irt, rep)
	})
	return irt
}

func getTags(sel *goquery.Selection) []string {
	var tags []string
	sel.Find(".p-category").Each(func(i int, s *goquery.Selection) {
		tag := strings.TrimPrefix(s.Text(), "#")
		tags = append(tags, tag)
	})
	return tags
}

func getFeaturedImage(sel *goquery.Selection) string {
	var img string
	sel.Find("meta").Each(func(i int, s *goquery.Selection) {
		v, _ := s.Attr("property")
		if v == "og:image" {
			img, _ = s.Attr("content")
		}
	})
	return img
}

func getDefaultImage(url string) string {
	d, err := getPage(url)
	if err != nil {
		return ""
	}
	s := d.Find("html")
	return getFeaturedImage(s)
}

func getTitle(sel *goquery.Selection) string {
	v := sel.Find(".idno-body").Find(".p-name").Find("a").Text()
	return v
}

func getPermalink(sel *goquery.Selection) string {
	v, _ := sel.Find(".permalink").Find(".u-url").Attr("href")
	v, _ = url.PathUnescape(v)
	return v
}

func getRelPermalink(sel *goquery.Selection) string {
	link := getPermalink(sel)
	u, err := url.Parse(link)
	if err != nil {
		// Known is buggy as hell
		uf := fixURL(link)
		u, err = url.Parse(uf)
		if err != nil {
			panic(err)
		}
	}
	return u.Path
}

func getDtPublished(sel *goquery.Selection) string {
	v, _ := sel.Find(".dt-published").Attr("datetime")
	return v
}

func getPage(uri string) (*goquery.Document, error) {
	res, err := http.Get(uri)
	if err != nil {
		// Known is buggy as hell
		u := fixURL(uri)
		res, err = http.Get(u)
		if err != nil {
			return nil, fmt.Errorf("can not parse URL")
		}
	}
	defer res.Body.Close()
	if res.StatusCode != 200 {
		err := fmt.Errorf("status code error: %d %s", res.StatusCode, res.Status)
		return nil, fmt.Errorf("can not get page: %w", err)
	}

	// Load the HTML document
	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		panic(err)
	}

	return doc, nil
}

func fixURL(uri string) string {
	pth := strings.TrimPrefix(uri, website)
	pth = strings.TrimPrefix(pth, "/")
	comps := strings.SplitN(pth, "/", 2)
	if len(comps) < 2 {
		return ""
	}
	parts := []string{website, comps[0], url.PathEscape(comps[1])}
	res := strings.Join(parts, "/")
	return res
}

func downloadFile(filepath string, url string) error {

	// Get the data
	resp, err := http.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	// Create the file
	out, err := os.Create(filepath)
	if err != nil {
		return err
	}
	defer out.Close()

	// Write the body to file
	_, err = io.Copy(out, resp.Body)
	return err
}

func downloadImages(path string, images map[string]string) {
	for fn, url := range images {
		filename := filepath.Join(path, fn)
		if err := downloadFile(filename, url); err != nil {
			fmt.Printf("failed to fetch image: %s - %v\n", url, err)
			b := []byte(url)
			if err := ioutil.WriteFile(filename, b, 0644); err != nil {
				fmt.Println(err)
			}
		}
	}
}
